code/datums/storage/storage.dm

```diff
@@ -0,0 +1,1027 @@
+	RegisterSignal(parent, ATOM_RECALCULATE_STORAGE_SPACE, PROC_REF(recalculate_storage_space))
+	RegisterSignals(parent, list(COMSIG_ITEM_EQUIPPED, COMSIG_ITEM_DROPPED), PROC_REF(update_verbs))
+	RegisterSignal(parent, COMSIG_ITEM_QUICK_EQUIP, PROC_REF(on_quick_equip_request))
+
+///Unregisters our signals from parent. Used when parent loses storage but is not destroyed
+/datum/storage/proc/unregister_storage_signals(atom/parent)
+	UnregisterSignal(parent, list(
+		COMSIG_ATOM_ATTACKBY,
+		COMSIG_ATOM_ATTACK_HAND,
+		COMSIG_ITEM_ATTACK_SELF,
+		COMSIG_ATOM_ATTACK_HAND_ALTERNATE,
+		COMSIG_CLICK_ALT,
+		COMSIG_CLICK_ALT_RIGHT,
+		COMSIG_CLICK_CTRL,
+		COMSIG_ATOM_ATTACK_GHOST,
+		COMSIG_MOUSEDROP_ONTO,
+
+		COMSIG_ATOM_EMP_ACT,
+		COMSIG_CONTENTS_EX_ACT,
+
+		COMSIG_ATOM_CONTENTS_DEL,
+		ATOM_MAX_STACK_MERGING,
+		ATOM_RECALCULATE_STORAGE_SPACE,
+		COMSIG_ITEM_EQUIPPED,
+		COMSIG_ITEM_DROPPED,
+		COMSIG_ITEM_QUICK_EQUIP,
+	))
+
+/// Almost 100% of the time the lists passed into set_holdable are reused for each instance
+/// Just fucking cache it 4head
+/// Yes I could generalize this, but I don't want anyone else using it. in fact, DO NOT COPY THIS
+/// If you find yourself needing this pattern, you're likely better off using static typecaches
+/// I'm not because I do not trust implementers of the storage component to use them, BUT
+/// IF I FIND YOU USING THIS PATTERN IN YOUR CODE I WILL BREAK YOU ACROSS MY KNEES
+/// ~Lemon
+GLOBAL_LIST_EMPTY(cached_storage_typecaches)
+
+/datum/storage/proc/set_holdable(list/can_hold_list, list/cant_hold_list, list/storage_type_limits_list)
+	if(!isnull(can_hold_list) && !islist(can_hold_list))
+		can_hold_list = list(can_hold_list)
+	if(!isnull(cant_hold_list) && !islist(cant_hold_list))
+		cant_hold_list = list(cant_hold_list)
+	if(!isnull(storage_type_limits_list) && !islist(storage_type_limits_list))
+		storage_type_limits_list = list(storage_type_limits_list)
+
+	if(!isnull(can_hold_list))
+		var/unique_key = can_hold_list.Join("-")
+		if(!GLOB.cached_storage_typecaches[unique_key])
+			GLOB.cached_storage_typecaches[unique_key] = typecacheof(can_hold_list)
+		can_hold = GLOB.cached_storage_typecaches[unique_key]
+
+	if(!isnull(cant_hold_list))
+		var/unique_key = cant_hold_list.Join("-")
+		if(!GLOB.cached_storage_typecaches[unique_key])
+			GLOB.cached_storage_typecaches[unique_key] = typecacheof(cant_hold_list)
+		cant_hold = GLOB.cached_storage_typecaches[unique_key]
+
+	if(!isnull(storage_type_limits_list))
+		var/unique_key = storage_type_limits_list.Join("-")
+		if(!GLOB.cached_storage_typecaches[unique_key])
+			GLOB.cached_storage_typecaches[unique_key] = typecacheof(storage_type_limits_list)
+		storage_type_limits = GLOB.cached_storage_typecaches[unique_key]
+
+///This proc is called when you want to place an attacking_item into the storage
+/datum/storage/proc/on_attackby(datum/source, obj/item/attacking_item, mob/user, params)
+	SIGNAL_HANDLER
+	if(length(refill_types))
```

Comment:
probably early return here

Reply:
Best I can do is at 315

Reply:
![image](https://github.com/tgstation/TerraGov-Marine-Corps/assets/66163761/aa586cf2-e4e6-4cba-9f8c-2185960fcd97)
But this looks a bit silly so I'd rather not tbh

Reply:
if(!length(refill_types))
		if(!can_be_inserted(attacking_item, user))
			return FALSE
		INVOKE_ASYNC(src, PROC_REF(handle_item_insertion), attacking_item, FALSE, user)
		return COMPONENT_NO_AFTERATTACK

	for(var/typepath in refill_types)
		if(istype(attacking_item, typepath))
			INVOKE_ASYNC(src, PROC_REF(do_refill), attacking_item, user)
			return
			
	It is not that much better though, so you chose

Reply:
Yeah this is just a lot more confusing to read, when the whole point in early returns is to reduce indentation and make code more readable, so I'd rather stick with current code.